File: r2k2zzn/k2zzn_dp/plugdp.js
"use strict";
/*
 * Plug DP feasibility solver for k=2 zig-zag numberlink on an R-row strip,
 * for any R >= 1.
 *
 * THE DATA
 * --------
 * A Plug describes what crosses one row of a vertical cut between two
 * columns. There are three kinds:
 *
 *   EMPTY    nothing crosses this row
 *   ANCHOR   a path fragment that has already touched ONE of its two
 *            terminals crosses here (it still needs to reach the other one)
 *   BRIDGE   a path fragment that has touched NEITHER terminal yet crosses
 *            here; it has exactly one other end, also a BRIDGE plug
 *            somewhere else in the same profile, carrying the same
 *            (color, fragment) pair
 *
 * A Profile is an array of R Plugs, one per row. `fragment` is renumbered
 * fresh every column (canonicalize) and only ever needs to distinguish
 * BRIDGE pairs open at the same time -- at most floor(R/2) of them.
 *
 * COST: exponential in R (profile-state count grows like a weighted
 * Motzkin number, ~5^R), linear in the number of columns.
 */

const EMPTY = { "kind": "EMPTY", "color": -1, "fragment": -1 };

function anchor(color) {
  return { "kind": "ANCHOR", "color": color, "fragment": -1 };
}
function bridge(color, fragment) {
  return { "kind": "BRIDGE", "color": color, "fragment": fragment };
}
function isEmpty(p) { return p.kind === "EMPTY"; }
function plugKey(p) {
  return p.kind + "," + p.color.toString() + "," + p.fragment.toString();
}
function profileKey(profile) {
  return profile.map(plugKey).join("|");
}

// Renumber BRIDGE fragment labels by order of first appearance.
//
function canonicalize(profile) {
  const remap = new Map();
  return profile.map((plug) => {
    if (plug.kind !== "BRIDGE") return plug;
    const key = plug.color + "," + plug.fragment;
    if (!remap.has(key)) { remap.set(key, remap.size); }
    return bridge(plug.color, remap.get(key));
  });
}

function terminalAt(terminals, row, col) {
  for (const t of terminals) {
    if ((t.row === row) && (t.col === col)) { return t.color; }
  }
  return null;
}

function findBridgePartner(working, excludeRow, color, fragment) {
  for (let row = 0; row < working.length; row++) {
    if (row === excludeRow) { continue; }
    const p = working[row];
    if ((p.kind === "BRIDGE") && (p.color === color) && (p.fragment === fragment)) { return row; }
  }
  return null;
}

function resolveCell(working, row, present, isTerminal, termColor, useRight, useDown, nextId) {
  const results = [];
  const place = (value) => {
    const w = working.slice();
    w[row] = useRight ? value : EMPTY;
    const p = useDown ? value : EMPTY;
    return [w, p, nextId];
  };

  if (present.length === 0) {
    if (isTerminal) {
      results.push(place(anchor(termColor)));
    } else {
      for (const color of [0, 1]) {
        const frag = bridge(color, nextId);
        const w = working.slice();
        w[row] = useRight ? frag : EMPTY;
        const p = useDown ? frag : EMPTY;
        results.push([w, p, nextId + 1]);
      }
    }
  } else if (present.length === 1) {
    const plug = present[0];
    if (!useRight && !useDown) {
      if (plug.color !== termColor) { return results; }
      if (plug.kind === "ANCHOR") {
        const w = working.slice();
        w[row] = EMPTY;
        results.push([w, EMPTY, nextId]);
      } else {
        const other = findBridgePartner(working, row, plug.color, plug.fragment);
        if (other === null) { return results; }
        const w = working.slice();
        w[other] = anchor(plug.color);
        w[row] = EMPTY;
        results.push([w, EMPTY, nextId]);
      }
    } else {
      results.push(place(plug));
    }
  } else {
    const [p1, p2] = present;
    if (p1.color !== p2.color) { return results; }
    const color = p1.color;
    if (p1.kind === "ANCHOR" && p2.kind === "ANCHOR") {
      const w = working.slice();
      w[row] = EMPTY;
      results.push([w, EMPTY, nextId]);
    } else if ((p1.kind === "ANCHOR") || (p2.kind === "ANCHOR")) {
      const fragPlug = ((p1.kind === "ANCHOR") ? p2 : p1);
      const other = findBridgePartner(working, row, color, fragPlug.fragment);
      if (other === null) { return results; }
      const w = working.slice();
      w[other] = anchor(color);
      w[row] = EMPTY;
      results.push([w, EMPTY, nextId]);
    } else {
      if (p1.fragment === p2.fragment) { return results; }
      const row1 = findBridgePartner(working, row, color, p1.fragment);
      const row2 = findBridgePartner(working, row, color, p2.fragment);
      if (row1 === null || row2 === null) { return results; }
      const w = working.slice();
      const newFrag = nextId;
      w[row1] = bridge(color, newFrag);
      w[row2] = bridge(color, newFrag);
      w[row] = EMPTY;
      results.push([w, EMPTY, nextId + 1]);
    }
  }
  return results;
}

function processColumn(profileIn, rows, col, numCols, terminals) {
  let freshId = 0;
  for (const p of profileIn) if (p.kind === "BRIDGE") freshId = Math.max(freshId, p.fragment + 1);

  let branches = [[profileIn.slice(), null, freshId]];

  for (let row = 0; row < rows; row++) {
    const nextBranches = [];
    for (const [working, pendingIn, nextId] of branches) {
      const left = col > 0 ? working[row] : "BOUNDARY";
      const up = row > 0 ? pendingIn : "BOUNDARY";
      const canRight = col < numCols - 1;
      const canDown = row < rows - 1;

      const termColor = terminalAt(terminals, row, col);
      const isTerminal = termColor !== null;
      const budget = isTerminal ? 1 : 2;

      const present = [left, up].filter((p) => p !== "BOUNDARY" && !isEmpty(p));
      const have = present.length;
      if (have > budget) { continue; }
      const need = budget - have;

      const available = [];
      if (canRight) available.push("right");
      if (canDown) available.push("down");
      if (need > available.length) continue;

      for (const choice of combinations(available, need)) {
        const useRight = choice.includes("right");
        const useDown = choice.includes("down");
        for (const out of resolveCell(working, row, present, isTerminal, termColor, useRight, useDown, nextId)) {
          nextBranches.push(out);
        }
      }
    }
    branches = nextBranches;
  }

  const seen = new Map();
  for (const [working] of branches) {
    const canon = canonicalize(working);
    seen.set(profileKey(canon), canon);
  }
  return [...seen.values()];
}

function combinations(arr, k) {

  //console.log(Array(k).fill("  ").join(""), "combo:", k, arr);

  if (k === 0) { return [[]]; }
  if (arr.length === 0) { return []; }
  const [first, ...rest] = arr;
  const withFirst = combinations(rest, k - 1).map((c) => [first, ...c]);
  const withoutFirst = combinations(rest, k);
  return withFirst.concat(withoutFirst);
}

function validateInput(rows, numCols, s0, t0, s1, t1) {
  // See the matching comment in plugdp.py: an out-of-bounds terminal is
  // invisible to the sweep, so that color is silently never required to
  // cover any cells, and the DP can wrongly report "feasible". Reject
  // this at the door rather than compute a misleading answer.
  if (rows < 1 || numCols < 1) {
    throw new Error(`rows and numCols must be >= 1, got rows=${rows}, numCols=${numCols}`);
  }
  const terminals = { s0, t0, s1, t1 };
  for (const [name, [r, c]] of Object.entries(terminals)) {
    if (!(r >= 0 && r < rows && c >= 0 && c < numCols)) {
      throw new Error(
        `${name}=(${r},${c}) is out of bounds for a ${rows}x${numCols} grid ` +
        `(rows must be in 0..${rows - 1}, cols in 0..${numCols - 1})`
      );
    }
  }
  const keys = new Set(Object.values(terminals).map(([r, c]) => r + "," + c));
  if (keys.size !== 4) {
    throw new Error(`s0, t0, s1, t1 must be four distinct cells, got ${JSON.stringify(terminals)}`);
  }
}

function isFeasible(rows, numCols, s0, t0, s1, t1) {
  validateInput(rows, numCols, s0, t0, s1, t1);
  const terminals = [
    { row: s0[0], col: s0[1], color: 0 },
    { row: t0[0], col: t0[1], color: 0 },
    { row: s1[0], col: s1[1], color: 1 },
    { row: t1[0], col: t1[1], color: 1 },
  ];
  let profiles = [Array(rows).fill(EMPTY)];
  for (let col = 0; col < numCols; col++) {
    const seen = new Map();
    for (const profile of profiles) {
      for (const next of processColumn(profile, rows, col, numCols, terminals)) {
        seen.set(profileKey(next), next);
      }
    }
    profiles = [...seen.values()];
    if (profiles.length === 0) return false;
  }
  const emptyKey = profileKey(Array(rows).fill(EMPTY));
  return profiles.some((p) => profileKey(p) === emptyKey);
}

// ---------------------------------------------------------------------------
// Path construction: not just "is it feasible" but "show me the two paths".
// See the matching comment in plugdp.py for why this needs a second pass
// with backpointers rather than reusing isFeasible's deduplicated profiles.
// ---------------------------------------------------------------------------

function processColumnTracked(profileIn, rows, col, numCols, terminals) {
  let freshId = 0;
  for (const p of profileIn) {
    if (p.kind === "BRIDGE") {
      freshId = Math.max(freshId, p.fragment + 1);
    }
  }

  let branches = [[profileIn.slice(), null, freshId, []]];

  for (let row = 0; row < rows; row++) {
    const nextBranches = [];
    for (const [working, pendingIn, nextId, edges] of branches) {
      const left = ((col > 0) ? working[row] : "BOUNDARY");
      const up = ((row > 0) ? pendingIn : "BOUNDARY");
      const canRight = (col < (numCols - 1));
      const canDown = (row < (rows - 1));

      const termColor = terminalAt(terminals, row, col);
      const isTerminal = (termColor !== null);
      const budget = (isTerminal ? 1 : 2);

      const present = [left, up].filter((p) => { return (p !== "BOUNDARY") && (!isEmpty(p)) });
      const have = present.length;
      if (have > budget) { continue; }
      const need = budget - have;

      const available = [];
      if (canRight) { available.push("right"); }
      if (canDown) { available.push("down"); }
      if (need > available.length) { continue; }

      let combos = combinations(available, need);
      console.log("combos[r=", row,"]{avail:", available,",need:", need,"}:", combos);

      //for (const choice of combinations(available, need)) {
      for (const choice of combos) {
        const useRight = choice.includes("right");
        const useDown = choice.includes("down");
        for (const out of resolveCell(working, row, present, isTerminal, termColor, useRight, useDown, nextId)) {
          const [w, p, id] = out;
          nextBranches.push([w, p, id, edges.concat([[useRight, useDown]])]);
        }
      }
    }
    branches = nextBranches;
  }

  // profileKey -> {profile, edges}
  //
  const result = new Map();
  for (const [working, , , edges] of branches) {
    const canon = canonicalize(working);
    const key = profileKey(canon);
    if (!result.has(key)) { result.set(key, { profile: canon, edges }); }
  }
  return result;
}

function findPath(rows, numCols, s0, t0, s1, t1) {
  validateInput(rows, numCols, s0, t0, s1, t1);
  const terminals = [
    { row: s0[0], col: s0[1], color: 0 },
    { row: t0[0], col: t0[1], color: 0 },
    { row: s1[0], col: s1[1], color: 1 },
    { row: t1[0], col: t1[1], color: 1 },
  ];
  const initial = Array(rows).fill(EMPTY);


  let profiles = new Map([[profileKey(initial), initial]]);

  // history[col]: Map destKey -> {sourceProfile, edges}
  //
  const history = [];

  for (let col = 0; col < numCols; col++) {

    console.log("\n\n---\ncol:", col);
    console.log(profiles);

    const nextMap = new Map();
    for (const profile of profiles.values()) {
      let CT = processColumnTracked(profile, rows, col, numCols, terminals);

      console.log("  ct:");
      console.log("  ", CT);

      //for (const [destKey, { profile: dest, edges }] of processColumnTracked(profile, rows, col, numCols, terminals)) {
      for (const [destKey, { "profile": dest, "edges": edges }] of CT) {

        console.log("    ", destKey, profile, edges, dest);

        if (!nextMap.has(destKey)) {
          nextMap.set(destKey, { "sourceProfile": profile, "edges": edges, "profile": dest });
        }

      }
    }

    history.push(nextMap);
    profiles = new Map([...nextMap.entries()].map(([k, v]) => [k, v.profile]));
    if (profiles.size === 0) return null;
  }

  console.log("colN:");
  console.log(profiles);

  const finalKey = profileKey(Array(rows).fill(EMPTY));
  if (!profiles.has(finalKey)) return null;

  const edgeGrids = new Array(numCols);
  let currentKey = finalKey;
  for (let col = numCols - 1; col >= 0; col--) {
    const entry = history[col].get(currentKey);
    edgeGrids[col] = entry.edges;
    currentKey = profileKey(entry.sourceProfile);
  }

  const rightUsed = new Set();
  const downUsed = new Set();
  for (let col = 0; col < numCols; col++) {
    for (let row = 0; row < rows; row++) {
      const [useRight, useDown] = edgeGrids[col][row];
      if (useRight) rightUsed.add(row + "," + col);
      if (useDown) downUsed.add(row + "," + col);
    }
  }

  function neighbors([r, c]) {
    const out = [];
    if (rightUsed.has(r + "," + c)) out.push([r, c + 1]);
    if (rightUsed.has(r + "," + (c - 1))) out.push([r, c - 1]);
    if (downUsed.has(r + "," + c)) out.push([r + 1, c]);
    if (downUsed.has((r - 1) + "," + c)) out.push([r - 1, c]);
    return out;
  }

  function trace(start, end) {
    const path = [start];
    let prev = null, cur = start;
    while (cur[0] !== end[0] || cur[1] !== end[1]) {
      const options = neighbors(cur).filter((n) => !prev || n[0] !== prev[0] || n[1] !== prev[1]);
      if (options.length === 0) return null; // would indicate a reconstruction bug
      prev = cur;
      cur = options[0];
      path.push(cur);
    }
    return path;
  }

  const path0 = trace(s0, t0);
  const path1 = trace(s1, t1);
  return [path0, path1];
}

function _print_instance(_row, _col, s0, t0, s1, t1) {
  let grid = [];
  for (let r=0; r<_row; r++) {
    let a = [];
    for (let c=0; c<_col; c++) {
      let ch = '.';
      if ((r == s0[0]) && (c == s0[1])) { ch='a'; }
      if ((r == t0[0]) && (c == t0[1])) { ch='A'; }
      if ((r == s1[0]) && (c == s1[1])) { ch='b'; }
      if ((r == t1[0]) && (c == t1[1])) { ch='B'; }
      a.push(ch);
    }
    grid.push(a);
  }

  for (let r=0; r<_row; r++) {
    console.log( grid[r].join("") );
  }
}

module.exports.findPath = findPath;
module.exports = { isFeasible, processColumn, EMPTY, canonicalize, profileKey };

if (require.main === module) {
  //console.log(isFeasible(3, 4, [0, 0], [1, 2], [0, 1], [2, 2]));
  //console.log(isFeasible(5, 6, [0, 0], [4, 5], [0, 5], [4, 0]));

  _print_instance(3, 4, [0, 0], [1, 2], [0, 1], [2, 2]);
  console.log(JSON.stringify(findPath(3, 4, [0, 0], [1, 2], [0, 1], [2, 2])));

  //_print_instance(3, 8, [0, 2], [2, 2], [0, 5], [2, 5]);
  //console.log(JSON.stringify(findPath(3, 8, [0, 2], [2, 2], [0, 5], [2, 5])));

}

File: r2k2zzn/k2zzn_dp/plugdp.py
"""
Plug DP feasibility solver for k=2 zig-zag numberlink on an R-row strip,
for any R >= 1.

THE DATA
--------
A `Plug` describes what crosses one row of a vertical cut between two
columns. There are three kinds:

    EMPTY    nothing crosses this row
    ANCHOR   a path fragment that has already touched ONE of its two
             terminals crosses here (it still needs to reach the other one)
    BRIDGE   a path fragment that has touched NEITHER terminal yet crosses
             here; it has exactly one other end, also a BRIDGE plug
             somewhere else in the same profile, carrying the same
             (color, fragment) pair

A `Profile` is a tuple of R Plugs, one per row -- the entire state of the
sweep at one vertical cut. `fragment` is not a stable identity: it is
renumbered fresh every column (see `canonicalize`) and only ever needs to
distinguish BRIDGE pairs that are open AT THE SAME TIME -- at most
floor(R/2) of them, by a simple packing argument (each pair occupies 2 of
the R rows). For R=3 that bound is 1, which is why a fixed-width-3 version
of this code never needs more than a single fragment label; for R>=4 it's
a real possibility and the code below handles it the same way regardless
of R, with no special-casing.

THE ALGORITHM
-------------
Sweep columns left to right. Within a column, process its R cells top to
bottom. At each cell, combine what arrives from the LEFT (previous column)
and from UP (the row above, within this same column) to decide what goes
RIGHT (next column) and DOWN (next row), according to the cell's budget:
1 open edge if it's one of the four given terminals, 2 otherwise.

COST
----
This is exponential in R (the profile-state count grows like a weighted
Motzkin number, roughly 5^R) but only LINEAR in the number of columns,
since the profile set is capped at that R-dependent constant no matter how
many columns are swept. Practical for R up to roughly a dozen; not a
substitute for a solution polynomial in both dimensions.
"""

from enum import Enum
from dataclasses import dataclass
from itertools import combinations


class Kind(Enum):
    EMPTY = "empty"
    ANCHOR = "anchor"
    BRIDGE = "bridge"


@dataclass(frozen=True)
class Plug:
    kind: Kind
    color: int = -1     # 0 or 1; meaningless when kind is EMPTY
    fragment: int = -1  # meaningless unless kind is BRIDGE (see module docstring)


EMPTY = Plug(Kind.EMPTY)


def anchor(color):
    return Plug(Kind.ANCHOR, color)


def bridge(color, fragment):
    return Plug(Kind.BRIDGE, color, fragment)


def canonicalize(profile):
    """Renumber BRIDGE fragment labels by order of first appearance, so
    equivalent connectivity patterns compare equal as dict/set keys."""
    remap = {}
    out = []
    for plug in profile:
        if plug.kind != Kind.BRIDGE:
            out.append(plug)
            continue
        key = (plug.color, plug.fragment)
        if key not in remap:
            remap[key] = len(remap)
        out.append(bridge(plug.color, remap[key]))
    return tuple(out)


@dataclass(frozen=True)
class Terminal:
    row: int
    col: int
    color: int


def find_bridge_partner(working, exclude_row, color, fragment):
    """The other row (not `exclude_row`) currently holding a BRIDGE plug
    with the same (color, fragment) -- i.e. this fragment's other end."""
    for row, plug in enumerate(working):
        if row == exclude_row:
            continue
        if plug.kind == Kind.BRIDGE and plug.color == color and plug.fragment == fragment:
            return row
    return None


def terminal_at(terminals, row, col):
    for t in terminals:
        if t.row == row and t.col == col:
            return t.color
    return None


def process_column(profile_in, rows, col, num_cols, terminals):
    """Process one column's `rows` cells top to bottom. Returns the set of
    canonicalized profiles reachable after finishing this column."""
    # Any freshly-minted fragment id must not collide with one already in
    # use by the incoming profile, or two unrelated fragments could be
    # confused for one -- see module docstring on `fragment`.
    used = [p.fragment for p in profile_in if p.kind == Kind.BRIDGE]
    fresh_fragment_start = max(used, default=-1) + 1

    branches = [(list(profile_in), None, fresh_fragment_start)]

    for row in range(rows):
        next_branches = []
        for working, pending_in, next_id in branches:
            left = working[row] if col > 0 else "BOUNDARY"
            up = pending_in if row > 0 else "BOUNDARY"
            can_go_right = col < num_cols - 1
            can_go_down = row < rows - 1

            term_color = terminal_at(terminals, row, col)
            is_terminal = term_color is not None
            budget = 1 if is_terminal else 2

            present = [p for p in (left, up) if p not in ("BOUNDARY", EMPTY)]
            have = len(present)
            if have > budget:
                continue  # over budget already -- dead end
            need = budget - have

            available = (["right"] if can_go_right else []) + (["down"] if can_go_down else [])
            if need > len(available):
                continue  # not enough room to reach budget -- dead end

            for choice in combinations(available, need):
                use_right = "right" in choice
                use_down = "down" in choice
                for out_working, out_pending, out_id in _resolve_cell(
                    working, row, present, have, is_terminal, term_color,
                    use_right, use_down, next_id,
                ):
                    next_branches.append((out_working, out_pending, out_id))
        branches = next_branches

    return {canonicalize(tuple(working)) for working, _, _ in branches}


def _resolve_cell(working, row, present, have, is_terminal, term_color,
                   use_right, use_down, next_id):
    """Yields (new_working, new_pending, new_next_id) for every legal way
    to resolve one cell, given what's already decided about its LEFT/UP
    edges (`present`, `have`) and which of RIGHT/DOWN are newly activated."""

    def place(value):
        w = list(working)
        w[row] = value if use_right else EMPTY
        p = value if use_down else EMPTY
        return w, p, next_id

    if have == 0:
        if is_terminal:
            # A terminal with no incoming edge starts a brand new ANCHOR.
            yield place(anchor(term_color))
        else:
            # An ordinary cell with no incoming edge starts a brand new
            # BRIDGE fragment. Its color isn't determined by anything seen
            # so far -- try both.
            for color in (0, 1):
                frag = bridge(color, next_id)
                w = list(working)
                w[row] = frag if use_right else EMPTY
                p = frag if use_down else EMPTY
                yield w, p, next_id + 1

    elif have == 1:
        plug = present[0]
        if not use_right and not use_down:
            # Terminal consuming its one incoming edge -- must match color.
            if plug.color != term_color:
                return
            if plug.kind == Kind.ANCHOR:
                w = list(working)
                w[row] = EMPTY
                yield w, EMPTY, next_id  # CLOSES this color's path
            else:
                other = find_bridge_partner(working, row, plug.color, plug.fragment)
                if other is None:
                    return
                w = list(working)
                w[other] = anchor(plug.color)  # ANCHOR the fragment's other end
                w[row] = EMPTY
                yield w, EMPTY, next_id
        else:
            # Ordinary cell: the incoming plug just passes straight through.
            yield place(plug)

    else:  # have == 2
        if use_right or use_down:
            return  # budget already met by LEFT+UP alone
        p1, p2 = present
        if p1.color != p2.color:
            return  # one cell can't serve two different paths
        color = p1.color
        if p1.kind == Kind.ANCHOR and p2.kind == Kind.ANCHOR:
            w = list(working)
            w[row] = EMPTY
            yield w, EMPTY, next_id  # two tips of one path meet: CLOSES it
        elif p1.kind == Kind.ANCHOR or p2.kind == Kind.ANCHOR:
            frag_plug = p2 if p1.kind == Kind.ANCHOR else p1
            other = find_bridge_partner(working, row, color, frag_plug.fragment)
            if other is None:
                return
            w = list(working)
            w[other] = anchor(color)
            w[row] = EMPTY
            yield w, EMPTY, next_id
        else:
            if p1.fragment == p2.fragment:
                return  # would close a loop touching no terminal -- illegal
            row1 = find_bridge_partner(working, row, color, p1.fragment)
            row2 = find_bridge_partner(working, row, color, p2.fragment)
            if row1 is None or row2 is None:
                return
            w = list(working)
            new_frag = next_id
            w[row1] = bridge(color, new_frag)
            w[row2] = bridge(color, new_frag)
            w[row] = EMPTY
            yield w, EMPTY, next_id + 1


def _validate_input(rows, num_cols, s0, t0, s1, t1):
    """Raises ValueError for anything that would silently produce a wrong
    answer rather than a clean failure: out-of-bounds terminals, or
    terminals that aren't four distinct cells. (Reported bug: an
    out-of-bounds terminal is invisible to the sweep -- terminal_at never
    matches it -- so that whole color is silently never required to cover
    any cells, and the DP can report "feasible" using only the other
    color's path across the entire grid. That's wrong, not just unusual,
    so it must be rejected here rather than computed.)"""
    if rows < 1 or num_cols < 1:
        raise ValueError(f"rows and num_cols must be >= 1, got rows={rows}, num_cols={num_cols}")
    terminals = {"s0": s0, "t0": t0, "s1": s1, "t1": t1}
    for name, (r, c) in terminals.items():
        if not (0 <= r < rows and 0 <= c < num_cols):
            raise ValueError(
                f"{name}={(r, c)} is out of bounds for a {rows}x{num_cols} grid "
                f"(rows must be in 0..{rows - 1}, cols in 0..{num_cols - 1})"
            )
    if len({s0, t0, s1, t1}) != 4:
        raise ValueError(f"s0, t0, s1, t1 must be four distinct cells, got {terminals}")


def is_feasible(rows, num_cols, s0, t0, s1, t1):
    """rows >= 1. s0, t0, s1, t1 are (row, col) pairs, row in 0..rows-1."""
    _validate_input(rows, num_cols, s0, t0, s1, t1)
    terminals = [
        Terminal(s0[0], s0[1], 0), Terminal(t0[0], t0[1], 0),
        Terminal(s1[0], s1[1], 1), Terminal(t1[0], t1[1], 1),
    ]
    profiles = {(EMPTY,) * rows}
    for col in range(num_cols):
        next_profiles = set()
        for profile in profiles:
            next_profiles |= process_column(profile, rows, col, num_cols, terminals)
        profiles = next_profiles
        if not profiles:
            return False
    return (EMPTY,) * rows in profiles



# ---------------------------------------------------------------------------
# Path construction: not just "is it feasible" but "show me the two paths".
#
# The feasibility DP above deliberately throws away history -- profiles get
# deduplicated, so by design you can't tell which of several ways of
# reaching a profile actually happened. To reconstruct an explicit witness,
# we re-run the same sweep, but this time record, for every profile reached
# at every column, ONE (source profile, per-row edge choices) pair that
# produced it -- enough to walk backward afterward and recover the full set
# of grid edges used, then trace the two paths out of that edge set.
# ---------------------------------------------------------------------------

def _process_column_tracked(profile_in, rows, col, num_cols, terminals):
    """Like process_column, but also returns, for each resulting profile,
    the per-row (use_right, use_down) choices for THIS column that produced
    it -- one witness per profile, not all of them."""
    used = [p.fragment for p in profile_in if p.kind == Kind.BRIDGE]
    fresh_fragment_start = max(used, default=-1) + 1

    # branch: (working, pending, next_id, edges-so-far for rows done in this column)
    branches = [(list(profile_in), None, fresh_fragment_start, [])]

    for row in range(rows):
        next_branches = []
        for working, pending_in, next_id, edges in branches:
            left = working[row] if col > 0 else "BOUNDARY"
            up = pending_in if row > 0 else "BOUNDARY"
            can_go_right = col < num_cols - 1
            can_go_down = row < rows - 1

            term_color = terminal_at(terminals, row, col)
            is_terminal = term_color is not None
            budget = 1 if is_terminal else 2

            present = [p for p in (left, up) if p not in ("BOUNDARY", EMPTY)]
            have = len(present)
            if have > budget:
                continue
            need = budget - have

            available = (["right"] if can_go_right else []) + (["down"] if can_go_down else [])
            if need > len(available):
                continue

            for choice in combinations(available, need):
                use_right = "right" in choice
                use_down = "down" in choice
                for out_working, out_pending, out_id in _resolve_cell(
                    working, row, present, have, is_terminal, term_color,
                    use_right, use_down, next_id,
                ):
                    next_branches.append((out_working, out_pending, out_id, edges + [(use_right, use_down)]))
        branches = next_branches

    result = {}
    for working, _, _, edges in branches:
        canon = canonicalize(tuple(working))
        if canon not in result:  # keep the first witness found; any one will do
            result[canon] = edges
    return result


def find_path(rows, num_cols, s0, t0, s1, t1):
    """Like is_feasible, but on success returns (path0, path1) -- each an
    explicit list of (row, col) cells from its start terminal to its end
    terminal, covering every cell of the grid between the two paths with
    no overlap. Returns None if infeasible."""
    _validate_input(rows, num_cols, s0, t0, s1, t1)
    terminals = [
        Terminal(s0[0], s0[1], 0), Terminal(t0[0], t0[1], 0),
        Terminal(s1[0], s1[1], 1), Terminal(t1[0], t1[1], 1),
    ]
    initial = (EMPTY,) * rows
    profiles = {initial}
    history = []  # history[col]: dest_profile -> (source_profile, edge_record)

    for col in range(num_cols):
        next_map = {}
        for profile in profiles:
            for dest, edges in _process_column_tracked(profile, rows, col, num_cols, terminals).items():
                if dest not in next_map:
                    next_map[dest] = (profile, edges)
        history.append(next_map)
        profiles = set(next_map.keys())
        if not profiles:
            return None

    final = (EMPTY,) * rows
    if final not in profiles:
        return None

    # Walk backward through the history to recover which edges were used
    # in every column of the actual accepted sweep.
    edge_grids = [None] * num_cols
    current = final
    for col in range(num_cols - 1, -1, -1):
        source, edges = history[col][current]
        edge_grids[col] = edges
        current = source

    right_used = set()  # (r, c) means edge (r, c) - (r, c+1) is used
    down_used = set()   # (r, c) means edge (r, c) - (r+1, c) is used
    for col in range(num_cols):
        for row in range(rows):
            use_right, use_down = edge_grids[col][row]
            if use_right:
                right_used.add((row, col))
            if use_down:
                down_used.add((row, col))

    def neighbors(cell):
        r, c = cell
        out = []
        if (r, c) in right_used:
            out.append((r, c + 1))
        if (r, c - 1) in right_used:
            out.append((r, c - 1))
        if (r, c) in down_used:
            out.append((r + 1, c))
        if (r - 1, c) in down_used:
            out.append((r - 1, c))
        return out

    def trace(start, end):
        path = [start]
        prev, cur = None, start
        while cur != end:
            options = [n for n in neighbors(cur) if n != prev]
            if not options:
                return None  # would indicate a reconstruction bug
            prev, cur = cur, options[0]
            path.append(cur)
        return path

    path0 = trace(s0, t0)
    path1 = trace(s1, t1)
    return path0, path1

if __name__ == "__main__":
    #print(is_feasible(3, 4, (0, 0), (1, 2), (0, 1), (2, 2)))   # True
    #print(is_feasible(5, 6, (0, 0), (4, 5), (0, 5), (4, 0)))   # R=5 example
    #print(find_path(3, 4, (0, 0), (1, 2), (0, 1), (2, 2)))     # explicit witness
    print(find_path(10, 10, (1, 1), (9, 9), (0, 9), (9, 0)))     # explicit witness
